File: leetcode_go/question/_1601_1700/_1694/sol_1694.go
// Package _1694 重新格式化电话号码
// @author goinhn
// @data 2022/10/07
package _1694

import "strings"

// 模拟
// time:O(n)
// space:O(n)
func reformatNumber(number string) string {
	s := strings.ReplaceAll(number, " ", "")
	s = strings.ReplaceAll(s, "-", "")

	var ans []string
	i := 0
	for ; i+4 < len(s); i += 3 {
		ans = append(ans, s[i:i+3])
	}

	s = s[i:]
	if len(s) < 4 {
		ans = append(ans, s)
	} else {
		ans = append(ans, s[:2], s[2:])
	}

	return strings.Join(ans, "-")
}

File: leetcode_java/java/com/goinhn/thorwcode/leetcode/question/_1601_1700/Sol1694.java
package com.goinhn.thorwcode.leetcode.question._1601_1700;

import java.util.ArrayList;
import java.util.List;

/**
 * 1694.重新格式化电话号码
 *
 * @author goinhn
 * @date 2022/10/1
 */
public class Sol1694 {

    /**
     * time:O(n)
     * space:O(n)
     */
    public String reformatNumber(String number) {
        List<Character> numList = new ArrayList<>();
        for (char ch : number.toCharArray()) {
            if (ch == ' ' || ch == '-') {
                continue;
            }
            numList.add(ch);
        }

        int limit = numList.size() % 3;
        if (limit == 1) {
            limit = 4;
        }

        if (limit == 0) {
            limit = 3;
        }

        StringBuilder str = new StringBuilder();
        int preLimit = numList.size() - limit;
        for (int i = 0; i < preLimit; i++) {
            str.append(numList.get(i));
            if ((i + 1) % 3 == 0) {
                str.append('-');
            }
        }

        if (limit == 3) {
            str.append(numList.get(preLimit))
                    .append(numList.get(preLimit + 1))
                    .append(numList.get(preLimit + 2));
        } else if (limit == 2) {
            str.append(numList.get(preLimit))
                    .append(numList.get(preLimit + 1));
        } else if (limit == 4) {
            str.append(numList.get(preLimit))
                    .append(numList.get(preLimit + 1))
                    .append('-')
                    .append(numList.get(preLimit + 2))
                    .append(numList.get(preLimit + 3));
        }

        return str.toString();
    }

}
